core/src/main/scala/com/evolutiongaming/kafka/flow/PartitionFlow.scala

```diff
@@ -283,9 +287,31 @@ object PartitionFlow {
     }
 
     val release: F[Unit] = if (config.commitOnRevoke) {
-      offsetToCommit.flatMap { offset =>
-        offset.traverse_ { offset =>
-          log.info(s"committing on revoke: $offset") *> scheduleCommit.schedule(offset)
+
+      // Outside F is about collecting handles to the offsets held by existing cache entries;
+      // Inside F is about calling and aggregating those handles, which can be done even after cache was cleared;
+      def offsetsHeldByCurrentKeys: F[F[Option[Offset]]] =
```

Comment:
please correct me if i'm wrong but does it make any difference if we calculate the min offset right away without having `F[F]`? both arms of the `traverse` below end up in `pure` and even if the value isn't ready yet it will be detected during the execution of the outer `F`, not the inner one. Is there any other point of making this effect nested and delaying executing it? If not, we could calculate these offsets right away and pass them with `.pure` to `offsetToCommit`

Reply:
> both arms of the traverse below end up in pure
Not exactly, I the first branch the outer F is `pure`, and the inner one capture the `holding` calculation, while in the second, the outer F performs logging, and the inner one is pure none, since we don't have anything to calculate.

Reply:
indeed, you're right! now i wonder, what is the point of not calculating these offsets beforehand? we ignore keys that are not yet ready in the outer computation, so clearing the cache shouldn't change anything in that regard

Reply:
Clearing the cache triggers release on `TimerFlow`s, which perform `flushOnCancel` if configured, which in turn releases the pending offset holds. Currently, we release the cache _after_ performing the commits, so `TimerFlow`s perform the persistence flush successfully, but by that time nobody cares about the fact that the offsets can be committed, so there is a de-sync between persistence and committed offsets on the next app start.